src/bin/cargo/cli.rs

```diff
@@ -222,16 +219,30 @@ fn add_ssl(version_string: &mut String) {
     }
 }
 
+/// Expands aliases recursively to collect all the command line arguments.
+///
+/// [`GlobalArgs`] need to be extracted before expanding aliases because the
+/// clap code for extracting a subcommand discards global options
+/// (appearing before the subcommand).
 fn expand_aliases(
     config: &mut Config,
     args: ArgMatches,
     mut already_expanded: Vec<String>,
 ) -> Result<(ArgMatches, GlobalArgs), CliError> {
     if let Some((cmd, args)) = args.subcommand() {
-        match (
-            commands::builtin_exec(cmd),
-            super::aliased_command(config, cmd)?,
-        ) {
+        let exec = commands::builtin_exec(cmd);
+        let aliased_cmd = super::aliased_command(config, cmd).or_else(|e| {
+            // HACK: `cargo version` must always work even with malformed configs.
+            // Here we treat the error as the aliased command not found,
+            // so it falls back to call builtin `cargo version`
+            if cmd == "version" {
```

Comment:
nit: Personally, I would prefer this in a separate PR as it is a distinct solution for a distinct root cause while for the shared error case

Reply:
https://github.com/rust-lang/cargo/blob/34cba461f295b9daf451c11b6452b3a62f398617/tests/testsuite/version.rs#L24-L28

There is already a test exercising the behaviour. That's the reason I did it in this PR.

Reply:
So your PR originally broke that test and needed to include this to keep the test working?

Reply:
Exactly!

Reply:
I'm thinking my PR edit will help clear this up for readers.

From

> In addition, cargo version must work even with 

To

> Despite this change, cargo version must still work even with

- "In addition' made me think this was a separate change
- "still" is trying to emphasize that it was working before